### ISSUE TITLE
Support OpenGL via RemoteFX for package validation.

### ISSUE THREAD

**Issue by denelon** [microsoft/winget-pkgs#12906]
# Description of the new feature/enhancement

Add support for packages with a dependency on OpenGL.
Some packages require OpenGL. The validation environment does not currently support this as a dependency. 

# Proposed technical implementation details (optional)

Once support for package dependencies is enabled, the validation environment should be able to test packages with the OpenGL dependencies via RemoteFX.

Examples:
https://github.com/microsoft/winget-pkgs/pull/705
https://github.com/microsoft/winget-pkgs/pull/1067

**Comment by MarNicGit** [comment]
Thanks for the effort of making this issue, @denelon, but I just went to check out RemoteFX and apparently when I used it a year ago for #1067 it was already unsupported being deprecated - [And it has been fully removed in the April security update of this year](https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/deploy/deploy-graphics-devices-using-remotefx-vgpu).

> Because of security concerns, RemoteFX vGPU is disabled by default on all versions of Windows starting with the July 14, 2020 Security Update and **removed** starting with the April 13, 2021 Security Update.

I'm gonna look into getting the alternative, [Discrete Device Assignment (DDA)](https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/deploy/deploying-graphics-devices-using-dda) working under Windows 10, though I'm not quite sure if it's possible at all sinds the docs only talk about Windows Server 2016 and up. 

If that's not going to work, well, then I don't know how we can possibly validate manifests for programs that require OpenGL inside the validation environment - the standard display driver simply doesn't cut it.

**Comment by denelon** [comment]
I guess I need to update the Issue. Thank you so much for digging into this!

**Comment by denelon** [comment]
We also have an interesting challenge with some device drivers requiring specific hardware to be present.

**Comment by MarNicGit** [comment]
@denelon I've done a bunch of digging and from what I understand DDA is not available on desktop versions of Windows 10 - You can only use it on server 2016 and up.

The only solutions currently available to consumers are qemu and VirtualBox, as far as I am aware. Naturally this is far from ideal: it's hard if not impossible to migrate the win10 dev environment to these platforms (and probably against the license agreement) and I doubt you guys would want to built this into the pipeline.

Do you think you could ask the Hyper-V team if there is currently any way of getting opengl in Hyper-v to work on consumer editions of Windows?

**Comment by jedieaston** [comment]
It's worth noting that Windows Sandbox supports graphics acceleration (see [here](https://docs.microsoft.com/en-us/windows/security/threat-protection/windows-sandbox/windows-sandbox-architecture#wddm-gpu-virtualization) for more info), so if you need a local environment to test packages that need OpenGL, that would work. There's a lot of limitations (one of these days I'm going to buy a ticket to Build so I can go ask the engineers responsible why I can't have more than one Sandbox open at a time), but it works. Tools\SandboxTest.ps1 is a good way to validate a package.

**Comment by denelon** [comment]
@jenatali and I just had a talk, this might be one option for getting openGL support in a VM:
https://devblogs.microsoft.com/directx/announcing-the-opencl-and-opengl-compatibility-pack-for-windows-10-on-arm/

**Comment by jenatali** [comment]
Yeah, right now you'd need the insider version of that package installed, but then it will enable OpenGL support on top of WARP (Windows Advanced Rasterization Platform), which is our D3D software renderer used in VMs with no hardware acceleration.

**Comment by jedieaston** [comment]
I did some more digging into this, turns out apparently vGPU support is just broken in Windows 10 20H2/21H1. This probably isn't a option, but OpenGL (and DirectX too, as I tried a game with DirectX on 21H1 and it didn't work) is available and works fine in Sandboxes created on build 21382.1000 of Windows 10 (possibly older versions, I enrolled my system into Insider Fast and was given that build):


Does anyone know if this is intended behavior? I think vGPU not working in Sandbox on 21H1 is a bug, but I don't know what to do other then send it into the Feedback Hub black hole.

**Comment by denelon** [comment]
If you don't mind, send a report to Feedback Hub and share the link here. I'll see if I can route it to the right team.

**Comment by jedieaston** [comment]
Whoop, [there it is.](https://aka.ms/AAcm0hu)

**Comment by pbozzay** [comment]
Hey @jedieaston - I work on Sandbox. Thanks for investigating; this does seem like a regression. Did you submit a repro trace by chance? I don't see traces attached to that FBH report, but if we were able to gather them it would help corroborate my current theory.

**Comment by jedieaston** [comment]
I did two recordings, although I don't know if they went through -- let me try again.

**Comment by jedieaston** [comment]
@pbozzay I just tried it from another system and it looks like Feedback Hub sent the data, let me know if you can't see it and I can try something else.

**Comment by pbozzay** [comment]
Thanks, I see more data now. Can you tell me what you get when you go here, on both devices (working/broken)?
http://aka.ms/gpucheck

**Comment by jedieaston** [comment]
Stable (21H1): 
![Capture](https://user-images.githubusercontent.com/21368066/119038799-8b110400-b981-11eb-8a2a-215fc4f09217.PNG)

Insider Fast (21382): 

![Capture2](https://user-images.githubusercontent.com/21368066/119038832-98c68980-b981-11eb-94d8-66a32568b095.PNG)

**Comment by pbozzay** [comment]
Hi @jedieaston  - I think I see the issue. I believe you may have accidentally put the <vGPU> tag inside of the LogonCommand node; the XML parser simply ignores this setting in this case. On current builds we enable vGPU by default, which is why it's still working for you there. 

This is what I see from your feedback:
![image](https://user-images.githubusercontent.com/1088800/119372737-8348ac80-bc6c-11eb-93c6-af8b86c88bdd.png)

Valid config will use vGPU in its own node:
```
<!-- 21H1 -->
<Configuration>
  <vGPU>Enable</vGPU>
  <LogonCommand> 
  ... 
  </LogonCommand>
</Configuration>
```

Now, one issue here is that we added functionality called VGPU "vendor extensions" after 21H1. OpenGL may rely on that functionality. Vendor extensions are enabled by default in newer builds (e.g. Insider Fast), or you can explicitly enable them like this:

```
<!-- Insider Fast build -->
<Configuration>
  <vGPU>EnableVendorExtensions</vGPU>
  <LogonCommand> 
  ... 
  </LogonCommand>
</Configuration>

```

Enabling vendor extensions may be the difference between OpenGL working or not, which may mean that on 21H1 Sandbox doesn't support OpenGL.

**Comment by MarNicGit** [comment]
@pbozzay Having updated to 21H1 earlier last week, I just went to try sandbox with the `vGPU` setting and we're getting somewhere!
![image](https://user-images.githubusercontent.com/47538428/119386729-398daf80-bcc8-11eb-9cfe-65c90593d845.png)

dxdiag also shows Direct3D working. ***However***, OpenGL still doesn't seem to work and crashes applications trying to use it. I've tried the following:

- Furmark, Super Tux, Osu!Lazer 2021.5240
  - Doesn't start, logs an error to eventlog
- OpenGL Extensions Viewer
  - Shows directx12 but crashes during opengl detection
- GfxBench
  - Silent crash during device detection
- Unigine Heaven Benchmark 4.0 basic edition
  - Runs in DX11 and DX9 mode but silently crashes when trying to run in OpenGL mode
    - Interestingly, the benchmark claims to run at 60 to 70fps in dx11 while it is clearly stuttering all over the place. Probably unrelated though.

Do you know if it is possible to install the insider build without switching my entire windows install to insider?

**Comment by denelon** [comment]
> Do you know if it is possible to install the insider build without switching my entire windows install to insider?

I don't know, I'll see if I can find out.

**Comment by denelon** [comment]
Maybe you could create a VM with an insider build, but I don't know if the Sandbox works in that kind of scenario.

**Comment by jedieaston** [comment]
@pbozzay D'oh! Thanks for the help.

> Maybe you could create a VM with an insider build, but I don't know if the Sandbox works in that kind of scenario.

This works but you need to enable [nested virtualization](https://docs.microsoft.com/en-us/virtualization/hyper-v-on-windows/user-guide/nested-virtualization) in Hyper-V so Windows Sandbox can be enabled. And then since you can't pass through a GPU, I don't know if you'd get acceleration or not.

**Comment by pbozzay** [comment]
@MarNicGit @jedieaston that's good progress. Unfortunately in this case I suspect this won't work without upgrading to a new insider build that has VendorExtensions. I believe if you join an insider preview build, you can then revert back to 21H1 and retain your personal files/programs (admittedly this isn't ideal).

**Comment by MarNicGit** [comment]
@pbozzay Thanks for the heads up, I'll look into switching to insider later tomorrow after watching MS Build.

**Comment by pbozzay** [comment]
My first guess is that enabling vendor extensions hasn't "changed" anything for some reason. I've reached out to the vGPU owners inside of Microsoft and will let you know what they think.

**Comment by MarNicGit** [comment]
@jedieaston I just noticed you had working OpenGL on build 21382, could you check if this is still the case when using the following config?
```
<Configuration>
    <vGPU>EnableVendorExtensions</vGPU>
</Configuration>
```

**Comment by jedieaston** [comment]
OpenGL confirmed when using that exact configuration on 21382. 

![image](https://user-images.githubusercontent.com/21368066/120479387-e678c400-c37b-11eb-8e33-bea7c17c2041.png)

Maybe AMD doesn't have a vendor extension.

**Comment by MarNicGit** [comment]
@jedieaston Thanks! That should help narrow things down a lot.

**Comment by RoadToDream** [comment]
This has to be included in the Sandbox documents.... Even explicitly enable vGPU in the configuration, OpenGL will not work properly.
